src/Data/WorldPeace/Union.hs

```diff
@@ -141,17 +141,26 @@ type NoElementError (r :: k) (rs :: [k]) =
 
 -- | This type family checks whether @a@ is inside @as@ and produces
 -- compile-time error if not.
-type family CheckElem (a :: k) (as :: [k]) :: Constraint where
-    CheckElem a as = MemberElem (Elem a as) a as
+type family CheckElemIsMember (a :: k) (as :: [k]) :: Constraint where
+    CheckElemIsMember a as =
+      If (Elem a as) (() :: Constraint) (TypeError (NoElementError a as))
 
--- | Helper type family. There is no lazy @if@ on type level, so we need to
--- pattern on the result of the 'Elem' type family to decide what do we need to
--- return: empty constraint or 'TypeError'.
-type family MemberElem (isElem :: Bool) (a :: k) (as :: [k]) :: Constraint where
-    MemberElem 'True  a as = ()
-    MemberElem 'False a as = TypeError (NoElementError a as)
+-- | Type-level @if@.
+--
+-- >>> Refl :: If 'True String Double :~: String
+-- Refl
+-- >>> Refl :: If 'False String Double :~: Double
+-- Refl
+type family If (bool :: Bool) (thenCase :: k) (elseCase :: k) :: k where
+  If 'True thenCase _ = thenCase
+  If 'False _ elseCase = elseCase
```

Comment:
If this actually works then this is indeed nice, no need to implement additional type families. You can use `If` from `base` however:

* https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Type-Bool.html#t:If

Reply:
Oh, I didn't realize `If` was in base.  I'll use that then.

Reply:
I made this change here:

https://github.com/cdepillabout/world-peace/commit/ceb254962538374c7e4e84b446214c756164006a

Reply:
@cdepillabout Nice! Type errors are still not that common, so best-practices around using them are not yet established. I'm working on a library to make the process of writing the type errors messages easier, but we all can expect some suboptimal solutions for quite some time... Though, it's already pretty nice 🙂